adapters/au_sa.js

```diff
@@ -0,0 +1,148 @@
+    }
+    // Wrap everything in a try/catch in case something goes wrong
+    try {
+      // Format the data
+      var data = formatData(body, source);
+      if (data === undefined) {
+        return cb({message: 'Failure to parse data.'});
+      }
+      cb(null, data);
+    } catch (e) {
+      console.error(e);
+      return cb({message: 'Unknown adapter error.'});
+    }
+  });
+};
+
+// site locations retrieved from https://data.sa.gov.au/data/dataset/a768c1f5-9714-4576-90bd-9dddaaa66ce4
+var siteLocations = {
+  'chr': [ 138.4951966, -35.1349443 ],
+  'eli': [ 138.6957631, -34.6984929 ],
+  'ken': [ 138.6650977, -34.9214075 ],
+  'lef1': [ 138.4963475, -34.8386688 ],
+  'net': [ 138.549098, -34.9438035 ],
+  'nor': [ 138.6229313, -34.8620143 ],
+  'ptp_o': [ 138.0198974, -33.1947947 ],
+  'why_s': [ 137.5332255, -33.023596 ],
+  'lef2': [ 138.4978642, -34.79128 ],
+  'cbd': [ 138.6010841, -34.92889 ],
+  'ptp_t': [ 138.0037294, -33.1711633 ],
+  'why_w': [ 137.5860979, -33.0361164 ],
+  'pta': [ 137.7868467, -32.5100065 ]
+};
+
+// remove non numeric values
+function parseValue (value) {
+  if (value === null || value === 'NM' || value === 'NA') {
+    return null;
+  }
+
+  var number = Number(value);
+  if (Number.isNaN(number)) {
+    return null;
+  }
+
+  return number;
+}
+
+var formatData = function (data, source) {
+  var units = {
+    'no2': 'ppm',
+    'o3': 'ppm',
+    'co': 'ppm',
+    'so2': 'ppm',
+    'pm10': 'µg/m³',
+    'pm25': 'µg/m³'
+  };
+
+  var measurements = [];
+
+  // parse the csv feed, exclude # lines
+  var rows = parse(data, {
+    trim: true,
+    comment: '#',
+    relax_column_count: true
+  });
+
+  // header row contains the date/time
+  var day = rows[0][1];
+  var month = rows[0][2];
+  var year = rows[0][3];
+  var time = rows[0][4];
+
+  // according to https://data.sa.gov.au/data/dataset/recent-air-quality/resource/d8abf079-9c51-4a0c-b827-dca926c4e95b
+  // "Times shown ... are Australian Central Standard Time (ACST). During
+  // daylight savings an hour will need to be added to the times shown."
+  // Hence we specifiy the date time is in +09:30 ie. ACST, this then gets
+  // correctly formatted to local time in Australia/Adelaide
+  var date = moment.tz(`${day} ${month} ${year} ${time} +09:30`, 'DD MMMM YYYY HH:mm ZZ', 'Australia/Adelaide');
+  var dateObject = {utc: date.toDate(), local: date.format()};
+
+  // loop through the remaining csv rows, which each contain a location
+  for (var i = 1; i < rows.length; i++) {
+    var row = rows[i];
+
+    var siteName = row[2];
+
+    var siteRef = row[3].replace(/_dm.jpg$/, '');
+
+    var paramaterValues = {
+      'o3': parseValue(row[4]),
+      'co': parseValue(row[5]),
+      'no2': parseValue(row[6]),
+      'so2': parseValue(row[7]),
+      'pm10': parseValue(row[8]),
+      'pm25': parseValue(row[9])
+    };
+
+    // base measurement properties
+    var baseMeasurement = {
+      location: siteName,
+      city: 'Adelaide',
```

Comment:
The only only way this could be done is using a hardcoded lookup, is that okay?
For example in the source we have "Country,Spencer Gulf,Whyalla Walls St".

Country here means something like in the countryside so probably not usefull.
Spencer Gulf is like a higher level region.
Whyalla is the city/town
Walls St is the specific site, but sometimes the site name is just the city/town eg. "Port Augusta".

Reply:
Since it shouldn't change, I think hardcoded lookup is fine

Reply:
Sure I've hardcoded them now. For some reasons when I run `> node index.js --dryrun --verbose --source 'Australia - South Australia'` not all the results are listed, but the count is correct.